components/script/dom/underlyingsourcecontainer.rs

```diff
@@ -0,0 +1,225 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::ptr;
+use std::rc::Rc;
+
+use dom_struct::dom_struct;
+use js::jsapi::{Heap, IsPromiseObject, JSObject};
+use js::jsval::JSVal;
+use js::rust::{Handle as SafeHandle, HandleObject, HandleValue as SafeHandleValue, IntoHandle};
+
+use crate::dom::bindings::callback::ExceptionHandling;
+use crate::dom::bindings::codegen::Bindings::UnderlyingSourceBinding::UnderlyingSource as JsUnderlyingSource;
+use crate::dom::bindings::import::module::Error;
+use crate::dom::bindings::import::module::UnionTypes::ReadableStreamDefaultControllerOrReadableByteStreamController as Controller;
+use crate::dom::bindings::reflector::{reflect_dom_object_with_proto, DomObject, Reflector};
+use crate::dom::bindings::root::{Dom, DomRoot};
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::promise::Promise;
+use crate::dom::teeunderlyingsource::TeeUnderlyingSource;
+use crate::script_runtime::CanGc;
+
+/// <https://streams.spec.whatwg.org/#underlying-source-api>
+/// The `Js` variant corresponds to
+/// the JavaScript object representing the underlying source.
+/// The other variants are native sources in Rust.
+#[derive(JSTraceable)]
+#[crown::unrooted_must_root_lint::must_root]
+pub enum UnderlyingSourceType {
+    /// Facilitate partial integration with sources
+    /// that are currently read into memory.
+    Memory(usize),
+    /// A blob as underlying source, with a known total size.
+    Blob(usize),
+    /// A fetch response as underlying source.
+    FetchResponse,
+    /// A struct representing a JS object as underlying source,
+    /// and the actual JS object for use as `thisArg` in callbacks.
+    Js(JsUnderlyingSource, Heap<*mut JSObject>),
+    /// Tee
+    Tee(Dom<TeeUnderlyingSource>),
+}
+
+impl UnderlyingSourceType {
+    /// Is the source backed by a Rust native source?
+    pub fn is_native(&self) -> bool {
+        matches!(
+            self,
+            UnderlyingSourceType::Memory(_) |
+                UnderlyingSourceType::Blob(_) |
+                UnderlyingSourceType::FetchResponse
+        )
+    }
+
+    /// Does the source have all data in memory?
+    pub fn in_memory(&self) -> bool {
+        matches!(self, UnderlyingSourceType::Memory(_))
+    }
+}
+
+/// Wrapper around the underlying source.
+/// Useful because `Call_` requires the "this object" to impl DomObject.
```

Comment:
this second sentence is incorrect now and can be removed, see the doc at https://github.com/servo/servo/pull/34064/files#diff-6824c2479f7c9b0a016aa571d09fd9a01dc1dc7f442c13630e2d75d0578537d3R39